File: pkg/planner/core/operator/logicalop/logical_lock.go
// Copyright 2016 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package logicalop

import (
	"github.com/pingcap/tidb/pkg/expression"
	"github.com/pingcap/tidb/pkg/parser/ast"
	"github.com/pingcap/tidb/pkg/planner/core/base"
	"github.com/pingcap/tidb/pkg/planner/property"
	"github.com/pingcap/tidb/pkg/planner/util"
	"github.com/pingcap/tidb/pkg/planner/util/optimizetrace"
	"github.com/pingcap/tidb/pkg/planner/util/utilfuncp"
	"github.com/pingcap/tidb/pkg/util/plancodec"
)

// LogicalLock represents a select lock plan.
type LogicalLock struct {
	BaseLogicalPlan

	Lock         *ast.SelectLockInfo
	TblID2Handle map[int64][]util.HandleCols

	// tblID2phyTblIDCol is used for partitioned tables,
	// the child executor need to return an extra column containing
	// the Physical Table ID (i.e. from which partition the row came from)
	TblID2PhysTblIDCol map[int64]*expression.Column
}

// Init initializes LogicalLock.
func (p LogicalLock) Init(ctx base.PlanContext) *LogicalLock {
	p.BaseLogicalPlan = NewBaseLogicalPlan(ctx, plancodec.TypeLock, &p, 0)
	return &p
}

// *************************** start implementation of logicalPlan interface ***************************

// HashCode inherits BaseLogicalPlan.LogicalPlan.<0th> implementation.

// PredicatePushDown inherits BaseLogicalPlan.LogicalPlan.<1st> implementation.

// PruneColumns implements base.LogicalPlan.<2nd> interface.
func (p *LogicalLock) PruneColumns(parentUsedCols []*expression.Column, opt *optimizetrace.LogicalOptimizeOp) (base.LogicalPlan, error) {
	var err error
	if !IsSelectForUpdateLockType(p.Lock.LockType) {
		// when use .baseLogicalPlan to call the PruneColumns, it means current plan itself has
		// nothing to pruning or plan change, so they resort to its children's column pruning logic.
		// so for the returned logical plan here, p is definitely determined, we just need to collect
		// those extra deeper call error in handling children's column pruning.
		_, err = p.BaseLogicalPlan.PruneColumns(parentUsedCols, opt)
		if err != nil {
			return nil, err
		}
		return p, nil
	}

	for tblID, cols := range p.TblID2Handle {
		for _, col := range cols {
			for i := 0; i < col.NumCols(); i++ {
				parentUsedCols = append(parentUsedCols, col.GetCol(i))
			}
		}
		if physTblIDCol, ok := p.TblID2PhysTblIDCol[tblID]; ok {
			// If the children include partitioned tables, there is an extra partition ID column.
			parentUsedCols = append(parentUsedCols, physTblIDCol)
		}
	}
	p.Children()[0], err = p.Children()[0].PruneColumns(parentUsedCols, opt)
	if err != nil {
		return nil, err
	}
	return p, nil
}

// FindBestTask inherits BaseLogicalPlan.LogicalPlan.<3rd> implementation.

// BuildKeyInfo inherits BaseLogicalPlan.LogicalPlan.<4th> implementation.

// PushDownTopN implements the base.LogicalPlan.<5th> interface.
func (p *LogicalLock) PushDownTopN(topNLogicalPlan base.LogicalPlan, opt *optimizetrace.LogicalOptimizeOp) base.LogicalPlan {
	var topN *LogicalTopN
	if topNLogicalPlan != nil {
		topN = topNLogicalPlan.(*LogicalTopN)
	}
	if topN != nil {
		p.Children()[0] = p.Children()[0].PushDownTopN(topN, opt)
	}
	return p.Self()
}

// DeriveTopN inherits BaseLogicalPlan.LogicalPlan.<6th> implementation.

// PredicateSimplification inherits BaseLogicalPlan.LogicalPlan.<7th> implementation.

// ConstantPropagation inherits BaseLogicalPlan.LogicalPlan.<8th> implementation.

// PullUpConstantPredicates inherits BaseLogicalPlan.LogicalPlan.<9th> implementation.

// RecursiveDeriveStats inherits BaseLogicalPlan.LogicalPlan.<10th> implementation.

// DeriveStats inherits BaseLogicalPlan.LogicalPlan.<11th> implementation.

// ExtractColGroups inherits BaseLogicalPlan.LogicalPlan.<12th> implementation.

// PreparePossibleProperties inherits BaseLogicalPlan.LogicalPlan.<13th> implementation.

// ExhaustPhysicalPlans implements base.LogicalPlan.<14th> interface.
func (p *LogicalLock) ExhaustPhysicalPlans(prop *property.PhysicalProperty) ([]base.PhysicalPlan, bool, error) {
	return utilfuncp.ExhaustPhysicalPlans4LogicalLock(p, prop)
}

// ExtractCorrelatedCols inherits BaseLogicalPlan.LogicalPlan.<15th> implementation.

// MaxOneRow inherits BaseLogicalPlan.LogicalPlan.<16th> implementation.

// Children inherits BaseLogicalPlan.LogicalPlan.<17th> implementation.

// SetChildren inherits BaseLogicalPlan.LogicalPlan.<18th> implementation.

// SetChild inherits BaseLogicalPlan.LogicalPlan.<19th> implementation.

// RollBackTaskMap inherits BaseLogicalPlan.LogicalPlan.<20th> implementation.

// CanPushToCop inherits BaseLogicalPlan.LogicalPlan.<21st> implementation.

// ExtractFD inherits BaseLogicalPlan.LogicalPlan.<22nd> implementation.

// GetBaseLogicalPlan inherits BaseLogicalPlan.LogicalPlan.<23rd> implementation.

// ConvertOuterToInnerJoin inherits BaseLogicalPlan.LogicalPlan.<24th> implementation.

// *************************** end implementation of logicalPlan interface ***************************

// IsSelectForUpdateLockType checks if the select lock type is for update type.
func IsSelectForUpdateLockType(lockType ast.SelectLockType) bool {
	if lockType == ast.SelectLockForUpdate ||
		lockType == ast.SelectLockForShare ||
		lockType == ast.SelectLockForUpdateNoWait ||
		lockType == ast.SelectLockForUpdateWaitN {
		return true
	}
	return false
}

File: rust/crates/tidb-planner/src/logical_lock.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Logical SELECT-lock classification from
//! `pkg/planner/core/operator/logicalop/logical_lock.go`.
//!
//! TiDB's AST owns the complete SelectLockType parser enum. This dependency-
//! closed leaf keeps its source iota discriminants and ports the logical plan's
//! three predicates: supported locks are the three FOR UPDATE variants and the
//! two FOR SHARE variants; skip-locked variants and unknown values remain
//! unsupported. AST metadata, column pruning, child plans, session context,
//! and lock planning/execution remain external boundaries.

/// Raw source-shaped `ast.SelectLockType` value.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub struct SelectLockType(i32);

impl SelectLockType {
    /// `ast.SelectLockNone`.
    pub const NONE: Self = Self(0);
    /// `ast.SelectLockForUpdate`.
    pub const FOR_UPDATE: Self = Self(1);
    /// `ast.SelectLockForShare`.
    pub const FOR_SHARE: Self = Self(2);
    /// `ast.SelectLockForUpdateNoWait`.
    pub const FOR_UPDATE_NO_WAIT: Self = Self(3);
    /// `ast.SelectLockForUpdateWaitN`.
    pub const FOR_UPDATE_WAIT_N: Self = Self(4);
    /// `ast.SelectLockForShareNoWait`.
    pub const FOR_SHARE_NO_WAIT: Self = Self(5);
    /// `ast.SelectLockForUpdateSkipLocked`.
    pub const FOR_UPDATE_SKIP_LOCKED: Self = Self(6);
    /// `ast.SelectLockForShareSkipLocked`.
    pub const FOR_SHARE_SKIP_LOCKED: Self = Self(7);

    /// Creates a lock type from an AST-compatible raw discriminant.
    #[must_use]
    pub const fn from_raw(raw: i32) -> Self {
        Self(raw)
    }

    /// Returns the source raw discriminant.
    #[must_use]
    pub const fn raw(self) -> i32 {
        self.0
    }
}

/// Returns whether the source lock type acquires a FOR UPDATE pessimistic lock.
#[must_use]
pub const fn is_for_update_lock_type(lock_type: SelectLockType) -> bool {
    let raw = lock_type.raw();
    raw == SelectLockType::FOR_UPDATE.raw()
        || raw == SelectLockType::FOR_UPDATE_NO_WAIT.raw()
        || raw == SelectLockType::FOR_UPDATE_WAIT_N.raw()
}

/// Returns whether the source lock type acquires a FOR SHARE pessimistic lock.
#[must_use]
pub const fn is_for_share_lock_type(lock_type: SelectLockType) -> bool {
    let raw = lock_type.raw();
    raw == SelectLockType::FOR_SHARE.raw() || raw == SelectLockType::FOR_SHARE_NO_WAIT.raw()
}

/// Returns whether TiDB supports acquiring a pessimistic lock for this type.
#[must_use]
pub const fn is_supported_select_lock_type(lock_type: SelectLockType) -> bool {
    is_for_update_lock_type(lock_type) || is_for_share_lock_type(lock_type)
}

#[cfg(test)]
mod tests {
    use super::{
        is_for_share_lock_type, is_for_update_lock_type, is_supported_select_lock_type,
        SelectLockType,
    };

    #[test]
    fn update_variants_match_source_predicate() {
        for lock_type in [
            SelectLockType::FOR_UPDATE,
            SelectLockType::FOR_UPDATE_NO_WAIT,
            SelectLockType::FOR_UPDATE_WAIT_N,
        ] {
            assert!(is_for_update_lock_type(lock_type));
            assert!(!is_for_share_lock_type(lock_type));
            assert!(is_supported_select_lock_type(lock_type));
        }
    }

    #[test]
    fn share_variants_match_source_predicate() {
        for lock_type in [SelectLockType::FOR_SHARE, SelectLockType::FOR_SHARE_NO_WAIT] {
            assert!(!is_for_update_lock_type(lock_type));
            assert!(is_for_share_lock_type(lock_type));
            assert!(is_supported_select_lock_type(lock_type));
        }
    }

    #[test]
    fn none_skip_locked_and_unknown_values_are_unsupported() {
        for raw in [
            SelectLockType::NONE.raw(),
            SelectLockType::FOR_UPDATE_SKIP_LOCKED.raw(),
            SelectLockType::FOR_SHARE_SKIP_LOCKED.raw(),
            -1,
            99,
        ] {
            let lock_type = SelectLockType::from_raw(raw);
            assert!(!is_for_update_lock_type(lock_type));
            assert!(!is_for_share_lock_type(lock_type));
            assert!(!is_supported_select_lock_type(lock_type));
        }
    }

    #[test]
    fn raw_discriminants_preserve_go_iota_order() {
        assert_eq!(SelectLockType::NONE.raw(), 0);
        assert_eq!(SelectLockType::FOR_UPDATE.raw(), 1);
        assert_eq!(SelectLockType::FOR_SHARE.raw(), 2);
        assert_eq!(SelectLockType::FOR_UPDATE_NO_WAIT.raw(), 3);
        assert_eq!(SelectLockType::FOR_UPDATE_WAIT_N.raw(), 4);
        assert_eq!(SelectLockType::FOR_SHARE_NO_WAIT.raw(), 5);
        assert_eq!(SelectLockType::FOR_UPDATE_SKIP_LOCKED.raw(), 6);
        assert_eq!(SelectLockType::FOR_SHARE_SKIP_LOCKED.raw(), 7);
    }
}
